File: benches/benchmark.js
/**
 * Benchmark for Mozilla Readability.js
 *
 * This script benchmarks the performance of Mozilla's Readability.js library
 * on the same test cases used by readabilityrs for fair comparison.
 */

import { Readability } from '@mozilla/readability';
import { JSDOM } from 'jsdom';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TEST_PAGES_DIR = path.join(__dirname, '..', 'tests', 'test-pages');

// Test cases grouped by size
// Note: wikipedia-2 is excluded due to excessive memory usage in JSDOM
const TEST_CASES = {
  small: ['001', '002', 'aclu'],
  medium: ['medium-1', 'nytimes-1', 'ars-1'],
  large: ['guardian-1', 'yahoo-2'],
  batch: ['001', '002', 'aclu', 'ars-1', 'bbc-1', 'buzzfeed-1', 'cnet', 'cnn', 'ehow-1', 'herald-sun-1']
};

function loadTestCase(name) {
  const filePath = path.join(TEST_PAGES_DIR, name, 'source.html');
  try {
    return fs.readFileSync(filePath, 'utf8');
  } catch (e) {
    return null;
  }
}

function parseWithReadability(html, url = 'https://example.com') {
  const dom = new JSDOM(html, { url });
  const reader = new Readability(dom.window.document);
  return reader.parse();
}

function benchmark(fn, iterations = 100) {
  // Warmup
  for (let i = 0; i < 3; i++) {
    fn();
  }

  // Force GC if available
  if (global.gc) {
    global.gc();
  }

  // Measure
  const times = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    const end = performance.now();
    times.push(end - start);
  }

  times.sort((a, b) => a - b);

  return {
    mean: times.reduce((a, b) => a + b, 0) / times.length,
    median: times[Math.floor(times.length / 2)],
    min: times[0],
    max: times[times.length - 1],
    p95: times[Math.floor(times.length * 0.95)],
    iterations
  };
}

function formatMs(ms) {
  if (ms < 1) {
    return `${(ms * 1000).toFixed(2)} µs`;
  }
  return `${ms.toFixed(2)} ms`;
}

function formatBytes(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
}

async function runBenchmarks() {
  console.log('='.repeat(70));
  console.log('Mozilla Readability.js Benchmark');
  console.log('='.repeat(70));
  console.log();

  const results = {
    single: {},
    batch: null,
    large: {}
  };

  // Single document benchmarks by size
  console.log('Single Document Parsing');
  console.log('-'.repeat(70));

  for (const [size, cases] of Object.entries(TEST_CASES)) {
    if (size === 'batch') continue;

    for (const testCase of cases) {
      const html = loadTestCase(testCase);
      if (!html) {
        console.log(`Skipping ${testCase}: file not found`);
        continue;
      }

      const iterations = size === 'large' ? 10 : 50;
      const result = benchmark(() => parseWithReadability(html), iterations);

      results.single[testCase] = {
        ...result,
        size: html.length,
        sizeCategory: size
      };

      console.log(`${testCase.padEnd(20)} (${formatBytes(html.length).padStart(10)}) | ` +
        `mean: ${formatMs(result.mean).padStart(12)} | ` +
        `median: ${formatMs(result.median).padStart(12)} | ` +
        `p95: ${formatMs(result.p95).padStart(12)}`);
    }
  }

  console.log();

  // Skip batch processing to avoid memory issues
  // The single document results are sufficient for comparison
  console.log('Batch Processing: Skipped (memory-intensive with JSDOM)');
  console.log();

  // Large documents are already included in single results above
  // Copy them to large category for comparison script compatibility
  for (const testCase of TEST_CASES.large) {
    if (results.single[testCase]) {
      results.large[testCase] = results.single[testCase];
    }
  }

  console.log('='.repeat(70));

  // Output JSON results for comparison script
  const outputPath = path.join(__dirname, 'js-results.json');
  fs.writeFileSync(outputPath, JSON.stringify(results, null, 2));
  console.log(`Results saved to: ${outputPath}`);
}

runBenchmarks().catch(console.error);

File: benches/benchmark.rs
//! Rust benchmark runner that outputs JSON results for comparison
//!
//! This can be run with: cargo run --release --example rust_benchmark
//! Or compiled and run directly.

use readabilityrs::Readability;
use serde::Serialize;
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::time::Instant;

#[derive(Serialize)]
struct BenchmarkResult {
    mean: f64,
    median: f64,
    min: f64,
    max: f64,
    p95: f64,
    iterations: usize,
    size: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    size_category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    document_count: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    total_size: Option<usize>,
}

#[derive(Serialize)]
struct Results {
    single: HashMap<String, BenchmarkResult>,
    batch: Option<BenchmarkResult>,
    large: HashMap<String, BenchmarkResult>,
}

fn load_test_case(name: &str) -> Option<String> {
    let path = Path::new("tests/test-pages").join(name).join("source.html");
    fs::read_to_string(&path).ok()
}

fn benchmark<F>(f: F, iterations: usize) -> BenchmarkResult
where
    F: Fn(),
{
    for _ in 0..5 {
        f();
    }

    let mut times: Vec<f64> = Vec::with_capacity(iterations);
    for _ in 0..iterations {
        let start = Instant::now();
        f();
        let elapsed = start.elapsed();
        times.push(elapsed.as_secs_f64() * 1000.0); // Convert to ms
    }

    times.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let mean = times.iter().sum::<f64>() / times.len() as f64;
    let median = times[times.len() / 2];
    let min = times[0];
    let max = times[times.len() - 1];
    let p95 = times[(times.len() as f64 * 0.95) as usize];

    BenchmarkResult {
        mean,
        median,
        min,
        max,
        p95,
        iterations,
        size: 0,
        size_category: None,
        document_count: None,
        total_size: None,
    }
}

fn format_ms(ms: f64) -> String {
    if ms < 1.0 {
        format!("{:.2} µs", ms * 1000.0)
    } else {
        format!("{:.2} ms", ms)
    }
}

fn format_bytes(bytes: usize) -> String {
    if bytes < 1024 {
        format!("{} B", bytes)
    } else if bytes < 1024 * 1024 {
        format!("{:.1} KB", bytes as f64 / 1024.0)
    } else {
        format!("{:.2} MB", bytes as f64 / (1024.0 * 1024.0))
    }
}

fn main() {
    // Note: wikipedia-2 is excluded due to its unusual DOM structure causing slow processing
    let test_cases: HashMap<&str, Vec<&str>> = [
        ("small", vec!["001", "002", "aclu"]),
        ("medium", vec!["medium-1", "nytimes-1", "ars-1"]),
        ("large", vec!["guardian-1", "yahoo-2"]),
    ]
    .into_iter()
    .collect();

    let batch_cases = vec![
        "001",
        "002",
        "aclu",
        "ars-1",
        "bbc-1",
        "buzzfeed-1",
        "cnet",
        "cnn",
        "ehow-1",
        "herald-sun-1",
    ];

    let mut results = Results {
        single: HashMap::new(),
        batch: None,
        large: HashMap::new(),
    };

    println!("{}", "=".repeat(70));
    println!("ReadabilityRS Benchmark (Rust)");
    println!("{}", "=".repeat(70));
    println!();

    println!("Single Document Parsing");
    println!("{}", "-".repeat(70));

    for (size_category, cases) in &test_cases {
        for test_case in cases {
            let html = match load_test_case(test_case) {
                Some(h) => h,
                None => {
                    println!("Skipping {}: file not found", test_case);
                    continue;
                }
            };

            let iterations = if *size_category == "large" { 20 } else { 100 };
            let html_clone = html.clone();

            let mut result = benchmark(
                || {
                    let readability = Readability::new(&html_clone, None, None).unwrap();
                    let _ = readability.parse();
                },
                iterations,
            );

            result.size = html.len();
            result.size_category = Some(size_category.to_string());

            println!(
                "{:20} ({:>10}) | mean: {:>12} | median: {:>12} | p95: {:>12}",
                test_case,
                format_bytes(html.len()),
                format_ms(result.mean),
                format_ms(result.median),
                format_ms(result.p95)
            );

            if *size_category == "large" {
                results.large.insert(test_case.to_string(), result);
            } else {
                results.single.insert(test_case.to_string(), result);
            }
        }
    }

    println!();

    println!("Batch Processing (10 documents)");
    println!("{}", "-".repeat(70));

    let batch_docs: Vec<String> = batch_cases
        .iter()
        .filter_map(|name| load_test_case(name))
        .collect();

    if !batch_docs.is_empty() {
        let total_size: usize = batch_docs.iter().map(|h| h.len()).sum();

        let mut result = benchmark(
            || {
                for html in &batch_docs {
                    let readability = Readability::new(html, None, None).unwrap();
                    let _ = readability.parse();
                }
            },
            50,
        );

        result.document_count = Some(batch_docs.len());
        result.total_size = Some(total_size);
        result.size = total_size;

        println!(
            "{} documents ({} total) | mean: {} | per-doc avg: {}",
            batch_docs.len(),
            format_bytes(total_size),
            format_ms(result.mean),
            format_ms(result.mean / batch_docs.len() as f64)
        );

        results.batch = Some(result);
    }

    println!();

    println!("Large Document Parsing");
    println!("{}", "-".repeat(70));

    for test_case in &["guardian-1", "yahoo-2"] {
        if let Some(result) = results.large.get(*test_case) {
            let throughput = result.size as f64 / result.mean / 1024.0;
            println!(
                "{:20} ({:>10}) | mean: {:>12} | throughput: {:.2} KB/ms",
                test_case,
                format_bytes(result.size),
                format_ms(result.mean),
                throughput
            );
        }
    }

    println!();
    println!("{}", "=".repeat(70));

    let json = serde_json::to_string_pretty(&results).unwrap();
    let output_path = "benchmark/rust-results.json";
    fs::write(output_path, &json).unwrap();
    println!("Results saved to: {}", output_path);
}
